In [2]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python Docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the read-only "../input/" directory
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        
        #print(os.path.join(dirname, filename))
        pass
    
# You can write up to 20GB to the current directory (/kaggle/working/) that gets preserved as output when you create a version using "Save & Run All" 
# You can also write temporary files to /kaggle/temp/, but they won't be saved outside of the current session

this note book isn't completed yet , so i work in my free time

In [3]:
#!pip install tensorflow

In [4]:
# to prevent cleanup message from tensorflow as there are alot of them appear
import tensorflow as tf

tf.get_logger().setLevel('INFO')

In [5]:
INPUT_DATA_PATH = "/kaggle/input/nft-art-dataset/"

# Loading The DataSet

In [6]:
data  = pd.read_csv(INPUT_DATA_PATH + "dataset/dataset.csv")

In [7]:
data.head()

title                  name         creator  \
0     30 min Drawings            Giant Frog      kristyglas   
1  Experimental Video          Biospecimens  juliakponsford   
2            Sexy Art             long legs         badsexy   
3         Dream World  A Guide in my Dreams         yoslehz   
4         Dream World       Silent Observer         yoslehz   

                                       art_series  price     symbol   type  \
0           kristyglas_30-min-drawings_giant-frog   50.0  SWAP.HIVE  PHOTO   
1  juliakponsford_experimental-video_biospecimens  500.0  SWAP.HIVE  VIDEO   
2                      badsexy_sexy-art_long-legs   10.0  SWAP.HIVE  PHOTO   
3        yoslehz_dream-world_a-guide-in-my-dreams   20.0  SWAP.HIVE  PHOTO   
4             yoslehz_dream-world_silent-observer   20.0  SWAP.HIVE    GIF   

   likes   nsfw  tokens  year  rights  royalty  \
0      2  False      30  2020       1        0   
1      0  False       1  2020       1        0   
2      0   True       2  2021       1        0   
3      1  False       2  2020       1        0   
4      0  False       2  2020       1        0   

                                              cid  \
0  QmegjZDN2gyD9A6cG4efj8DwbLLGTW6isDGadNC2yVJHfH   
1  QmeN4KuTQwHPWvGDhiopNJziqmf1feGeabDD4Gk7z9DgUL   
2  QmaifAjZTJd8jTxNrwSoGWmFpbPYcszMF4ZcZDBwpRR4v4   
3  QmUTicJApSyByFb7sgsNa6cz8MG5RKsHvGLP61gxSF3JsN   
4  Qmb8WbXXdorG52sgY4mSyYUjevE65SGa2vhSxdXstfinz1   

                                                path  
0  ./dataset/image/QmegjZDN2gyD9A6cG4efj8DwbLLGTW...  
1  ./dataset/video/QmeN4KuTQwHPWvGDhiopNJziqmf1fe...  
2  ./dataset/image/QmaifAjZTJd8jTxNrwSoGWmFpbPYcs...  
3  ./dataset/image/QmUTicJApSyByFb7sgsNa6cz8MG5RK...  
4  ./dataset/gif/Qmb8WbXXdorG52sgY4mSyYUjevE65SGa...

In [8]:
data['likes'].unique()

array([ 2,  0,  1,  3,  4,  5,  6,  9,  7, 10])

## Working on photo data not videos or gifs

In [9]:
PHOTO_DATA = data[data['type' ]== 'PHOTO']

In [10]:
PHOTO_DATA['type'].unique()

array(['PHOTO'], dtype=object)

In [11]:
PHOTO_DATA

title  \
0                                     30 min Drawings   
2                                            Sexy Art   
3                                         Dream World   
6                                         Dream World   
9                                        experimental   
...                                               ...   
4182                                     Mouse Fetish   
4183                                       The beauty   
4184                                 Abstract Erotica   
4186                                              Eye   
4188  fine arts -  high resolution printable artworks   

                                      name         creator  \
0                               Giant Frog      kristyglas   
2                                long legs         badsexy   
3                     A Guide in my Dreams         yoslehz   
6                 I found colors in my sky         yoslehz   
9                                 Angela 👼        elias15g   
...                                    ...             ...   
4182                   Marble Triple Click  rubenalexander   
4183                       Girl in garden.       alexa.art   
4184                       Taste of Heaven       tntdabomb   
4186                   Eye Of The Beholder  rubenalexander   
4188  Dirty laundry in the fractal kitchen          elgeko   

                                             art_series   price     symbol  \
0                 kristyglas_30-min-drawings_giant-frog   50.00  SWAP.HIVE   
2                            badsexy_sexy-art_long-legs   10.00  SWAP.HIVE   
3              yoslehz_dream-world_a-guide-in-my-dreams   20.00  SWAP.HIVE   
6          yoslehz_dream-world_i-found-colors-in-my-sky   20.00  SWAP.HIVE   
9                         elias15g_experimental_angela-   35.00  SWAP.HIVE   
...                                                 ...     ...        ...   
4182    rubenalexander_mouse-fetish_marble-triple-click  225.08  SWAP.HIVE   
4183                alexa.art_the-beauty_girl-in-garden   14.00  SWAP.HIVE   
4184         tntdabomb_abstract-erotica_taste-of-heaven   50.00  SWAP.HIVE   
4186             rubenalexander_eye_eye-of-the-beholder   50.00  SWAP.HIVE   
4188  elgeko_fine-arts-high-resolution-printable-art...  700.00  SWAP.HIVE   

       type  likes   nsfw  tokens  year  rights  royalty  \
0     PHOTO      2  False      30  2020       1        0   
2     PHOTO      0   True       2  2021       1        0   
3     PHOTO      1  False       2  2020       1        0   
6     PHOTO      0  False       2  2020       1        0   
9     PHOTO      3  False       5  2020       1        0   
...     ...    ...    ...     ...   ...     ...      ...   
4182  PHOTO      0  False       5  2020       1        0   
4183  PHOTO      0  False       7  2020       1        0   
4184  PHOTO      0  False      10  2020       3        0   
4186  PHOTO      0  False       3  2020       1        0   
4188  PHOTO      0  False       1  2020       1        0   

                                                 cid  \
0     QmegjZDN2gyD9A6cG4efj8DwbLLGTW6isDGadNC2yVJHfH   
2     QmaifAjZTJd8jTxNrwSoGWmFpbPYcszMF4ZcZDBwpRR4v4   
3     QmUTicJApSyByFb7sgsNa6cz8MG5RKsHvGLP61gxSF3JsN   
6     QmT36BPdeJWNEdUr566z4DYGWxXqrDSeoA7zCHbKw8528w   
9     QmWQGXNu47pecHGfEatd3ATakphq5zXVrzs5unYYK4VR1c   
...                                              ...   
4182  Qmbk1o7v6gxu9nVAPNfji1gWbQBT4JD9mkEKBLg46soEgP   
4183  QmYE2JsNxWbfqg3sYAfsxo84UkqBRixxTFmgBHLGSdbbhy   
4184  QmeNbzcYHKDzZFTuY3PqBPr5WGKMGmrjrTsKDKd7WTe422   
4186  QmWTT58erwH6Vtv8UJNNM6mmWuwZNsWAYygRXmHnpJeeBa   
4188  QmNYLy1vRaRYoGRMVmheRgMxPAcrUufoUJHFSe81DmzNB9   

                                                   path  
0     ./dataset/image/QmegjZDN2gyD9A6cG4efj8DwbLLGTW...  
2     ./dataset/image/QmaifAjZTJd8jTxNrwSoGWmFpbPYcs...  
3     ./dataset/image/QmUTicJApSyByFb7sgsNa6cz8MG5RK...  
6     ./dataset/image/QmT36BPdeJWNEdUr566z4DYGWxXqrD...

## clean the path feature by removing the first two chars -> ./ 
## as we will use the path laterly without the need for to chars -> ./


In [12]:
PHOTO_DATA['path'] = PHOTO_DATA['path'].apply(lambda x : x.replace('./', ''))

/opt/conda/lib/python3.7/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """Entry point for launching an IPython kernel.


In [13]:
PHOTO_DATA


title  \
0                                     30 min Drawings   
2                                            Sexy Art   
3                                         Dream World   
6                                         Dream World   
9                                        experimental   
...                                               ...   
4182                                     Mouse Fetish   
4183                                       The beauty   
4184                                 Abstract Erotica   
4186                                              Eye   
4188  fine arts -  high resolution printable artworks   

                                      name         creator  \
0                               Giant Frog      kristyglas   
2                                long legs         badsexy   
3                     A Guide in my Dreams         yoslehz   
6                 I found colors in my sky         yoslehz   
9                                 Angela 👼        elias15g   
...                                    ...             ...   
4182                   Marble Triple Click  rubenalexander   
4183                       Girl in garden.       alexa.art   
4184                       Taste of Heaven       tntdabomb   
4186                   Eye Of The Beholder  rubenalexander   
4188  Dirty laundry in the fractal kitchen          elgeko   

                                             art_series   price     symbol  \
0                 kristyglas_30-min-drawings_giant-frog   50.00  SWAP.HIVE   
2                            badsexy_sexy-art_long-legs   10.00  SWAP.HIVE   
3              yoslehz_dream-world_a-guide-in-my-dreams   20.00  SWAP.HIVE   
6          yoslehz_dream-world_i-found-colors-in-my-sky   20.00  SWAP.HIVE   
9                         elias15g_experimental_angela-   35.00  SWAP.HIVE   
...                                                 ...     ...        ...   
4182    rubenalexander_mouse-fetish_marble-triple-click  225.08  SWAP.HIVE   
4183                alexa.art_the-beauty_girl-in-garden   14.00  SWAP.HIVE   
4184         tntdabomb_abstract-erotica_taste-of-heaven   50.00  SWAP.HIVE   
4186             rubenalexander_eye_eye-of-the-beholder   50.00  SWAP.HIVE   
4188  elgeko_fine-arts-high-resolution-printable-art...  700.00  SWAP.HIVE   

       type  likes   nsfw  tokens  year  rights  royalty  \
0     PHOTO      2  False      30  2020       1        0   
2     PHOTO      0   True       2  2021       1        0   
3     PHOTO      1  False       2  2020       1        0   
6     PHOTO      0  False       2  2020       1        0   
9     PHOTO      3  False       5  2020       1        0   
...     ...    ...    ...     ...   ...     ...      ...   
4182  PHOTO      0  False       5  2020       1        0   
4183  PHOTO      0  False       7  2020       1        0   
4184  PHOTO      0  False      10  2020       3        0   
4186  PHOTO      0  False       3  2020       1        0   
4188  PHOTO      0  False       1  2020       1        0   

                                                 cid  \
0     QmegjZDN2gyD9A6cG4efj8DwbLLGTW6isDGadNC2yVJHfH   
2     QmaifAjZTJd8jTxNrwSoGWmFpbPYcszMF4ZcZDBwpRR4v4   
3     QmUTicJApSyByFb7sgsNa6cz8MG5RKsHvGLP61gxSF3JsN   
6     QmT36BPdeJWNEdUr566z4DYGWxXqrDSeoA7zCHbKw8528w   
9     QmWQGXNu47pecHGfEatd3ATakphq5zXVrzs5unYYK4VR1c   
...                                              ...   
4182  Qmbk1o7v6gxu9nVAPNfji1gWbQBT4JD9mkEKBLg46soEgP   
4183  QmYE2JsNxWbfqg3sYAfsxo84UkqBRixxTFmgBHLGSdbbhy   
4184  QmeNbzcYHKDzZFTuY3PqBPr5WGKMGmrjrTsKDKd7WTe422   
4186  QmWTT58erwH6Vtv8UJNNM6mmWuwZNsWAYygRXmHnpJeeBa   
4188  QmNYLy1vRaRYoGRMVmheRgMxPAcrUufoUJHFSe81DmzNB9   

                                                   path  
0     dataset/image/QmegjZDN2gyD9A6cG4efj8DwbLLGTW6i...  
2     dataset/image/QmaifAjZTJd8jTxNrwSoGWmFpbPYcszM...  
3     dataset/image/QmUTicJApSyByFb7sgsNa6cz8MG5RKsH...  
6     dataset/image/QmT36BPdeJWNEdUr566z4DYGWxXqrDSe...

### Show the likes values

In [14]:
likes = PHOTO_DATA['likes'].value_counts()

# remove the the likes values that associated with one NFT as it will be used in training 
# and no images remain for test
likes = likes[likes>1]
likes

0    2147
1     589
2     192
3      82
4      38
5      15
6       7
Name: likes, dtype: int64

### make a dictionary of likes

In [15]:
likes_stats = dict()
for i in range(len(likes)):
    likes_stats[likes.index[i]] = likes.values[i]
likes_stats


{0: 2147, 1: 589, 2: 192, 3: 82, 4: 38, 5: 15, 6: 7}

### make a list of likes files

In [16]:
likes_list = []
for a, b in enumerate(likes_stats.keys()):
    likes_list.append(b)
    
print(likes_list)
print(len(likes_list))

[0, 1, 2, 3, 4, 5, 6]
7


In [17]:
#data2= data.copy()

select likes and make a copied dataframe for them


In [18]:
#data2 = data2[data2['likes'].apply(lambda x : x in likes_list)]


In [19]:
#print(data2['likes'].unique())
#print(len(data2['likes'].unique()))

### so this is info of likes

In [20]:
#data2

In [21]:
#%cd /kaggle/working 
#%mkdir dataset
#%mkdir top5creator_images

In [22]:
%pwd

'/kaggle/working'

In [23]:
DDPATH = '/kaggle/working/'

In [24]:
len(os.listdir(INPUT_DATA_PATH + 'dataset/image'))

3021

In [25]:
#%rm -r top5creators/

## Create 10 Directories represent 10 likes values


In [26]:

os.makedirs('likes')
os.makedirs("likes_images")

In [27]:
%ls

__notebook_source__.ipynb  likes/  likes_images/


In [28]:
'''## Convert types of likes from numpy int to str as we can use it like a directory name 
for i in range(len(likes_list)):
    likes_list[i] = str(likes_list[i])
    
'''

'## Convert types of likes from numpy int to str as we can use it like a directory name \nfor i in range(len(likes_list)):\n    likes_list[i] = str(likes_list[i])\n    \n'

In [29]:
print(type(likes_list[i]))

<class 'numpy.int64'>


### create the sub dirs for each like value

In [30]:
likes_dir ='likes/'

In [31]:
for like_value in likes_list:
    os.makedirs(likes_dir + str(like_value))

In [32]:
%ls likes/

0/  1/  2/  3/  4/  5/  6/


## test the likes paths

In [33]:
PHOTO_DATA[PHOTO_DATA['likes' ]== likes_list[1]]['path']

3       dataset/image/QmUTicJApSyByFb7sgsNa6cz8MG5RKsH...
16      dataset/image/QmWSozMKfzNE43WoAmcf5QdL3ptrmBTP...
17      dataset/image/QmfDRZnYUtM96U8AgNbS9bqgpahvPSDr...
24      dataset/image/QmaDVBjsjmpPwHun3SCL9HGRjem5uZmo...
28      dataset/image/QmaM1zTWoiYwT5L4BNcaaRnB56ueaiXL...
                              ...                        
4035    dataset/image/QmWTwrhtMMCRFTbwAidwZEwCv7sMXtdk...
4039    dataset/image/QmXR4cwiNFdm7BKNdXvWVGF6PFK1uKRW...
4107    dataset/image/QmP3FMwXwxsPbatb4RyWCZjTtg8TTnp2...
4116    dataset/image/QmXR4cwiNFdm7BKNdXvWVGF6PFK1uKRW...
4158    dataset/image/QmVrUpVRnYHF99yrMZknr5o6bRmp9E8y...
Name: path, Length: 589, dtype: object

In [34]:
INPUT_DATA_PATH

'/kaggle/input/nft-art-dataset/'

In [35]:
%ls

__notebook_source__.ipynb  likes/  likes_images/


### copy images form dataset/images of each like value
### to thier new location in likes


In [36]:
# copy images of these creators to another images folder called top5_images
from tqdm import tqdm
import shutil
import random

i = 0
for like_value in likes_list:
    print(like_value)
    
    for path in tqdm(PHOTO_DATA[PHOTO_DATA['likes' ]== like_value]['path']):
        i+= 1
        

        shutil.copy(INPUT_DATA_PATH + path , likes_dir + str(like_value) ) 
    print(i)


0


100%|██████████| 2147/2147 [02:04<00:00, 17.31it/s]


2147
1


100%|██████████| 589/589 [00:30<00:00, 19.19it/s]


2736
2


100%|██████████| 192/192 [00:12<00:00, 15.35it/s]


2928
3


100%|██████████| 82/82 [00:05<00:00, 14.05it/s]


3010
4


100%|██████████| 38/38 [00:03<00:00, 11.89it/s]


3048
5


100%|██████████| 15/15 [00:01<00:00, 10.11it/s]


3063
6


100%|██████████| 7/7 [00:00<00:00, 11.21it/s]

3070


**Important note : you mean ask your self why the number of copied files is different from the number of images associated with each like value like in the likes_stats?**

**The answer is the values in the likes_stats is already correct but here is different**

**as an example : the likes_stats is = {0: 2950, 1: 781, 2: 270, 3: 109, 4: 49, 5: 19, 6: 7, 7: 2, 9: 1, 10: 1}**

**the like value 0: has 2950 NFTs associated this like value , but after coping the file you notice from above result that**

**the like value 0: has 2147  >>PHOTO<< NFTs notice the PHOTO Word here**

**So The Conclusion here is We Specified the data to be only photo so the other NFTs types :gifs and videos has also likes associated with them**

**But They had been exluded from the total number here as they not PHOTO type so the total number of likes will be decrease as we saw in the like_value : 10**

**In the likes_stats: 10 has one NFT associated with it , but now there is no NFTs in this directory as it may be video or gif so It had been excluded from acount**

**I hope you got it now**

**Now We must remove the directory that has no NFTs or has just one , as the model will not able to work on just image as train, so there will no images remain for test**

In [41]:
likes_stats

{0: 2147, 1: 589, 2: 192, 3: 82, 4: 38, 5: 15, 6: 7}

In [39]:
for like_value in likes_list:
    print(like_value)
    print(len(os.listdir(likes_dir + str(like_value))))

0
2103
1
587
2
191
3
82
4
37
5
14
6
7


In [ ]:
for like_value in likes_list:
    print(like_value)

In [38]:
for like in likes_list:
    
    print(len(os.listdir(likes_dir + str(like))))

2103
587
191
82
37
14
7


In [131]:
# test if we succuesfully make a folder for likes 
#and copy their images from images in the data set to the new directory for each one of them

%ls likes/9

### copy images form dataset/images of likes values
### to thier new location in likes_images
### and split them automatically to train val test split
### and each directory of train, val, test,
### has inside subdirs for each like value
### then the images

In [37]:
os.makedirs('likes_images/train')
os.makedirs('likes_images/test')
os.makedirs('likes_images/val')


In [45]:
classes_dir_str = []
for i in range(len(likes_list)):
    classes_dir_str.append(str(likes_list[i]))
    
classes_dir_str

['0', '1', '2', '3', '4', '5', '6']

In [46]:
import os
import numpy as np
import shutil
import random

root_dir = '/kaggle/working/' # data root path
classes_dir = classes_dir_str #total labels

val_ratio = 0.15
test_ratio = 0.05

for cls in classes_dir:
    
    
    if not os.path.isdir(root_dir + 'likes_images/train/' + cls):
        os.makedirs(root_dir + 'likes_images/train/' + cls)
        
    if not os.path.isdir(root_dir + 'likes_images/val/' + cls):
        os.makedirs(root_dir +'likes_images/val/' + cls)
        
    if not os.path.isdir(root_dir + 'likes_images/test/' + cls):
        os.makedirs(root_dir +'likes_images/test/' + cls)


    # Creating partitions of the data after shuffeling
    src = root_dir + 'likes/' + cls # Folder to copy images from
    #print("src -> " , src, "\n")

    allFileNames = os.listdir(src)
    #print("allFileNames -> " , allFileNames ,"\n")

    np.random.shuffle(allFileNames)

    train_FileNames, val_FileNames, test_FileNames = np.split(np.array(allFileNames),
                                                              [int(len(allFileNames)* (1 - (val_ratio + test_ratio))), 
                                                               int(len(allFileNames)* (1 - test_ratio))])

    #print("train_FileNames -> " , train_FileNames ,"\n")
    #print("val_FileNames -> " , val_FileNames ,"\n")
    #print("test_FileNames -> " , test_FileNames ,"\n")
    #print("==================================== \n")


    train_FileNames = [src+'/'+ name for name in train_FileNames.tolist()]
    #print("train_FileNames -> " , train_FileNames ,"\n")

    val_FileNames = [src+'/' + name for name in val_FileNames.tolist()]
    #print("val_FileNames -> " , val_FileNames ,"\n")

    test_FileNames = [src+'/' + name for name in test_FileNames.tolist()]
    #print("test_FileNames -> " , test_FileNames ,"\n")


    #print('Total images: ', len(allFileNames))
    #print('Training: ', len(train_FileNames))
    #print('Validation: ', len(val_FileNames))
    #print('Testing: ', len(test_FileNames))


    #print("*******************************************")
    # Copy-pasting images
    for name in train_FileNames:
        shutil.copy(name, root_dir + 'likes_images/' +'train/' + cls)
       
    #print("-----------------------------------------")
    for name in val_FileNames:
        shutil.copy(name, root_dir+ 'likes_images/' +'val/' + cls)
        
    #print("-----------------------------------------") 
    for name in test_FileNames:
        shutil.copy(name, root_dir+ 'likes_images/' +'test/' + cls)
       

OSError: [Errno 28] No space left on device

> make a test to make sure that the copied images is belong to the same like_value

> in the dataset.csv and our copy is sucssesfuly

> so just take the cid of the below image with out taking the extintion eg. .png , .jpg or .jpeg 

> just the cid

In [47]:
PHOTO_DATA[PHOTO_DATA['cid'] == 'QmRr5azUCkfUWb4V1BsbVYmkHtY4TGJYCydiqurPoo1TeN']

title                        name creator  \
2802  Printable NFT  Welcome to City, Muahah!!!    doze   

                                     art_series  price     symbol   type  \
2802  doze_printable-nft_welcome-to-city-muahah  120.0  SWAP.HIVE  PHOTO   

      likes   nsfw  tokens  year  rights  royalty  \
2802      0  False       2  2020       1        0   

                                                 cid  \
2802  QmRr5azUCkfUWb4V1BsbVYmkHtY4TGJYCydiqurPoo1TeN   

                                                   path  
2802  dataset/image/QmRr5azUCkfUWb4V1BsbVYmkHtY4TGJY...

> so after looking to like value we find our copy is succesfuly done

# Loading Data

In [ ]:
# make some configuration for batch_size , image_height, image_width 
# we make a low batch_size to save our memory from crashing

batch_size = 32
img_height = 180
img_width = 180

## Loading train images 

In [48]:
train_ds = tf.keras.utils.image_dataset_from_directory(
  '/kaggle/working/likes_images/train',
  seed=123,
  image_size=(img_height, img_width),
  batch_size=batch_size)

NameError: name 'img_height' is not defined

In [ ]:
train_ds

In [ ]:
class_names = train_ds.class_names
print(class_names)

## Loading valedation dataset 

In [ ]:
val_ds = tf.keras.utils.image_dataset_from_directory(
  '/kaggle/working/likes_images/val',
  seed=123,
  image_size=(img_height, img_width),
  batch_size=batch_size)

val_ds

## Loading test dataset

In [ ]:
test_ds = tf.keras.utils.image_dataset_from_directory(
  '/kaggle/working/likes_images/test',
  seed=123,
  image_size=(img_height, img_width),
  batch_size=batch_size)

test_ds

## display some examples of train images
#### there is a red message appear when we run the following code
#### but it is not a error or warning message ,
#### i think it is just an ordinary message form tensorflow

In [ ]:
tf.__version__




#### i think the clean up message is showen as there is a new version from tensorflow is 
#### more more updated from the current version of tensorflow on kaggle kernel 

In [ ]:
import matplotlib.pyplot as plt

plt.figure(figsize=(10, 10))
for images, labels in train_ds.take(1):
    for i in range(9):
        ax = plt.subplot(3, 3, i + 1)
        plt.imshow(images[i].numpy().astype("uint8"))
        plt.title(class_names[labels[i]])
        plt.axis("off")

## display some test images

In [ ]:
import matplotlib.pyplot as plt

plt.figure(figsize=(10, 10))
for images, labels in test_ds.take(1):
    for i in range(9):
        ax = plt.subplot(3, 3, i + 1)
        plt.imshow(images[i].numpy().astype("uint8"))
        plt.title(class_names[labels[i]])
        plt.axis("off")

## display val images 

In [ ]:
import matplotlib.pyplot as plt

plt.figure(figsize=(10, 10))
for images, labels in val_ds.take(1):
    print(type(images))
    for i in range(9):
        ax = plt.subplot(3, 3, i + 1)
        plt.imshow(images[i].numpy().astype("uint8"))
        plt.title(class_names[labels[i]])
        plt.axis("off")

In [ ]:
for image_batch, labels_batch in train_ds:
    print(image_batch.shape)
    print(labels_batch.shape)
    break

# Standardize the data

> The RGB channel values are in the [0, 255] range. This is not ideal for a neural network; in general you should seek to make your input values small.

> Here, you will standardize values to be in the [0, 1] range by using tf.keras.layers.Rescaling:

In [ ]:
normalization_layer = tf.keras.layers.Rescaling(1./255)


# Configure the dataset for performance 

In [ ]:
AUTOTUNE = tf.data.AUTOTUNE

train_ds = train_ds.cache().shuffle(1000).prefetch(buffer_size = AUTOTUNE)
val_ds = val_ds.cache().prefetch(buffer_size = AUTOTUNE)

# Create the model

In [ ]:
# import dependecieces to build our model 
from tensorflow import keras
from tensorflow.keras import layers
from tensorflow.keras.models import Sequential

In [ ]:
num_classes = len(class_names)

model = Sequential([
    layers.Rescaling(1./255, input_shape = (img_height, img_width, 3)),
    layers.Conv2D(16, 3, padding = 'same', activation = 'relu'),
    layers.MaxPooling2D(),
    layers.Conv2D(32, 3, padding = 'same' , activation = 'relu'),
    layers.MaxPooling2D(),
    layers.Conv2D(64, 3, padding = 'same', activation = 'relu'),
    layers.MaxPooling2D(),
    layers.Flatten(),
    layers.Dense(128, activation = 'relu'),
    layers.Dense(num_classes)
    
])

# Compile The Model 

> I used Adam optimizer 

> Sparse Categorical Cross Entropy as loss function 

> 

In [ ]:
model.compile(optimizer = 'adam',
             loss = tf.keras.losses.SparseCategoricalCrossentropy(from_logits = True),
             metrics = ['accuracy'])

# Model Summary

In [ ]:
model.summary()

# Train the model

In [ ]:
epochs = 10 
history = model.fit(train_ds,
                   validation_data = val_ds,
                   epochs = epochs)


# Visualize Trianing Results 

In [ ]:
acc = history.history['accuracy']
val_acc = history.history['val_accuracy']

loss = history.history['loss']
val_loss = history.history['val_loss']

epochs_range = range(epochs)

plt.figure(figsize = (8,8))
plt.subplot(1,2,1)
plt.plot(epochs_range, acc, label = 'Training Accuracy')
plt.plot(epochs_range, val_acc, label = 'Validation Accuracy')
plt.legend(loc = 'lower right')
plt.title('Training and Validation Accuracy')

plt.subplot(1,2,2)
plt.plot(epochs_range, loss, label = 'Training Loss')
plt.plot(epochs_range, val_loss, label = 'Validation Loss')
plt.legend(loc = 'upper right')
plt.title('Training and Validation Loss')
plt.show()

We can see in the above plots taht training accuracy is increasing like linearly over time,

but validation accuracy stalls around 65% (note : it varies from time to time when i restart the seesion on kaggle )in the training process , 

also the difference in accuracy between training and validation accuracy is noticable 

so this is a sign of overfitting 

and When there are a small number of training examples ,

the model sometimes learns from noises or unwanted details from trining examples ,

to extent that it negatively impacts the perfoemance of the model on new expamples .

To prevent overvitting I will use Dropout and data augmentation 



# Testing on This Model and Prediction

## We will take a test image and try to predict it

## for example we will take an iamge from elenasteem creator and see if the model can predict this creator or not ! :)

In [ ]:
%ls top5creators_images/test/elenasteem 

In [ ]:
img_path = 'top5creators_images/test/elenasteem/QmUAvkEDyzwEsb32B1QuEa49kUeH8iMvpfhgLAN1sTVYTt.png'

In [ ]:
# show the image
import matplotlib.image as mpimg

test_img = mpimg.imread(img_path)
plt.imshow(test_img)


In [ ]:
test_img = tf.keras.utils.load_img(img_path, target_size = (img_height, img_width))

img_array = tf.keras.utils.img_to_array(test_img)
img_array = tf.expand_dims(img_array, 0) # create the batch

predictions = model.predict(img_array)
score = tf.nn.softmax(predictions[0])


print(
"This image most likely belongs to {} with a {:.2f} percent confidence .".format(class_names[np.argmax(score)], 100 * np.max(score)))

## testing all images in elenasteem creator directory

In [ ]:
%ls top5creators_images/test/elenasteem

In [ ]:
test_imgs = tf.keras.utils.image_dataset_from_directory(
  'top5creators_images/test',
  seed=123,
  image_size=(img_height, img_width),
  batch_size=batch_size)

In [ ]:
# number of images in test folder 
i = 1

for images, labels in test_imgs.take(1):
    
    if i <21:
        print("labels = ", labels)
        #plt.imshow(mpimg.imread(images[i]))
        plt.imshow(images[i].numpy().astype("uint8"))
        img_array = tf.keras.utils.img_to_array(images[i])
        img_array = tf.expand_dims(img_array, 0) # create the batch

        predictions = model.predict(img_array)
        score = tf.nn.softmax(predictions[0])


        print(
        "This image most likely belongs to {} with a {:.2f} percent confidence .".format(class_names[np.argmax(score)], 100 * np.max(score)))
    i+=1

## to make sure that this images is belongs to the same creator in the prediction 
## i will print all images in test directory and thier labels

In [ ]:
plt.figure(figsize=(10, 10))
for images, labels in test_imgs.take(1):
    print(type(images))
    
    # we specify 21 according to number of images in the test directory to show all images 
    # and compare them with the predicted image
    for i in range(9):
        ax = plt.subplot(3, 3, i + 1)
        plt.imshow(images[i].numpy().astype("uint8"))
        plt.title(class_names[labels[i]])
        plt.axis("off")

# Overfitting 

#### We can see the availabel images in test directory it's varies every time we run this notebook

#### as the distribution of images in every time is randomlly dirstibuted from top5creators directory

#### to top5creators_images directory but we can make sure that this image belongs to the creator 

#### by exploring out csv file and select it's cid 

#### and we can copy the cid from test directory after we make usre that is belongs to

#### the same creator , and then assign this path to the following varibale img_path

# Data Augmentaion

In [ ]:
data_augmentation = keras.Sequential([
    layers.RandomFlip("horizontal",
                     input_shape=(img_height,img_width,3)),
    layers.RandomRotation(0.1),
    layers.RandomZoom(0.1)
])

## Visualize some random images after augmentated them

In [ ]:
plt.figure(figsize=(10,10))

for images, _ in train_ds.take(1):
    for i in range(9):
        augmented_images = data_augmentation(images)
        ax = plt.subplot(3, 3, i + 1)
        plt.imshow(augmented_images[0].numpy().astype("uint8"))
        plt.axis("off")

# Dropout

> When Using dropout to a lyer , it randomly dropouts a number of output units from the layer during the training process 

> Dropout takes a fractional number as it's input value , in the form such as 0.2 , 0.4 , and so on 

> this means dropping out by 20% or 40% of the output unit randomly from the applied layer

In [ ]:
model = Sequential([
    data_augmentation,
    layers.Rescaling(1./255),
    layers.Conv2D(16, 3, padding='same', activation='relu'),
    layers.MaxPooling2D(),
    layers.Conv2D(32, 3, padding='same', activation = 'relu'),
    layers.MaxPooling2D(),
    layers.Conv2D(64, 3, padding = 'same', activation = 'relu'),
    layers.MaxPooling2D(),
    layers.Dropout(0.2),
    layers.Flatten(),
    layers.Dense(128, activation = 'relu'),
    layers.Dense(num_classes)
])

## Compile and train the model

In [ ]:
model.compile(optimizer = 'adam',
             loss = tf.keras.losses.SparseCategoricalCrossentropy(from_logits = True),
             metrics = ['accuracy'])

In [ ]:
model.summary()

In [ ]:
epochs = 100

history = model.fit(
train_ds,
validation_data = val_ds,
epochs = epochs)

## Visualize training results 

In [ ]:
acc = history.history['accuracy']
val_acc = history.history['val_accuracy']

loss = history.history['loss']
val_loss = history.history['val_loss']

epochs_range = range(epochs)

plt.figure(figsize =  (8,8))
plt.subplot(1,2,1)
plt.plot(epochs_range, acc, label='Training Accuracy')
plt.plot(epochs_range, val_acc, label='Validation Accuracy')
plt.legend(loc = 'lower right')
plt.title('Training and Validation Acuuracy')

plt.subplot(1,2,2)
plt.plot(epochs_range, loss, label='Training Loss')
plt.plot(epochs_range, val_loss, label='Validation Loss')
plt.legend(loc = 'upper right')
plt.title('Training and Validation Loss')
plt.show()

# Testing The Model and Prediction

### We will take a test image and try to predict it

## for example we will take an iamge from elgeko creator and see if the model can predict this creator or not ! :)

In [ ]:
%ls top5creators_images/test/elgeko

In [ ]:
PHOTO_DATA[PHOTO_DATA['creator'] == 'elgeko']['cid']

In [ ]:
img_path = 'top5creators_images/test/elgeko/QmZqYLaEtqeCdAbSjJiMPDGoirxGmb6hEoXxmSWHTApRUC.png'

In [ ]:
# show the image
import matplotlib.image as mpimg

test_img = mpimg.imread(img_path)
plt.imshow(test_img)


In [ ]:
test_img = tf.keras.utils.load_img(img_path, target_size = (img_height, img_width))
img_array = tf.keras.utils.img_to_array(test_img)
img_array = tf.expand_dims(img_array, 0) # create the batch

predictions = model.predict(img_array)
score = tf.nn.softmax(predictions[0])


print(
"This image most likely belongs to {} with a {:.2f} percent confidence ."
.format(class_names[np.argmax(score)], 100 * np.max(score)))

## I will use another approach based on files locations 
## so i will use only the top5creators directory to load images 
## and split them to train and val sets

### loading data

In [ ]:
import pathlib
data_dir = pathlib.Path("top5creators")


### loading train, val and test datasets

In [ ]:
image_count = len(list(data_dir.glob('*/*.*')))

list_ds = tf.data.Dataset.list_files(str(data_dir/'*/*'), shuffle=False)
list_ds = list_ds.shuffle(image_count, reshuffle_each_iteration=False)


val_size = int(image_count * 0.2)
test_size = int(image_count * 0.1)

train_ds = list_ds.skip(val_size)
test_ds = list_ds.take(test_size)
val_ds = list_ds.take(val_size)


In [ ]:
image_count

In [ ]:
print(tf.data.experimental.cardinality(train_ds).numpy())
print(tf.data.experimental.cardinality(val_ds).numpy())
print(tf.data.experimental.cardinality(test_ds).numpy())


In [ ]:
print(list(range(10,20)))

In [ ]:
for f in train_ds.take(15):
    print(f.numpy())

In [ ]:
def get_label(file_path):
  # Convert the path to a list of path components
    parts = tf.strings.split(file_path, os.path.sep)
    print("parts = ", parts)
  # The second to last is the class-directory
    one_hot = parts[-2] == class_names
    print("parts[-2] = ", parts[-2])
    print("file_path = ", file_path)
    print("one hot = ",one_hot)
    print("class names = ", class_names)
    
  # Integer encode the label
    return tf.argmax(one_hot)

In [ ]:
def decode_img(img):
  # Convert the compressed string to a 3D uint8 tensor
  img = tf.io.decode_image(img, channels=3,expand_animations = False )
  # Resize the image to the desired size
  return tf.image.resize(img, [img_height, img_width])

In [ ]:
def process_path(file_path):
    label = get_label(file_path)
    # Load the raw data from the file as a string
    img = tf.io.read_file(file_path)
    img = decode_img(img)
    return img, label

In [ ]:
# Set `num_parallel_calls` so multiple images are loaded/processed in parallel.
train_ds = train_ds.map(process_path, num_parallel_calls=AUTOTUNE)
val_ds = val_ds.map(process_path, num_parallel_calls=AUTOTUNE)
test_ds = test_ds.map(process_path, num_parallel_calls = AUTOTUNE)

In [ ]:
print(tf.data.experimental.cardinality(train_ds).numpy())
print(tf.data.experimental.cardinality(val_ds).numpy())
print(tf.data.experimental.cardinality(test_ds).numpy())


In [ ]:
for image, label in train_ds.take(1):
      print("Image shape: ", image.numpy().shape)
      print("Label: ", label.numpy())

In [ ]:
for image, label in val_ds.take(1):
      print("Image shape: ", image.numpy().shape)
      print("Label: ", label.numpy())

In [ ]:
for image, label in test_ds.take(3):
    print("Image shape: ", image.numpy().shape)
    print("Label: ", label.numpy())
    #plt.imshow(image[0].numpy().astype("uint8"))  

In [ ]:
def configure_for_performance(ds):
    ds = ds.cache()
    ds = ds.shuffle(buffer_size=1000)
    #ds = ds.batch(batch_size)
    ds = ds.batch(5)
    ds = ds.prefetch(buffer_size=AUTOTUNE)
    return ds

train_ds = configure_for_performance(train_ds)
val_ds = configure_for_performance(val_ds)
test_ds = configure_for_performance(test_ds)

In [ ]:
#image_batch, label_batch = next(iter(test_ds))


'''
plt.figure(figsize=(10, 10))
for i in range(tf.data.experimental.cardinality(test_ds).numpy()):
    image, label in val_ds.take(i)
    ax = plt.subplot(3, 3, i + 1)
    plt.imshow(image[i].numpy().astype("uint8"))
    label = label[i]
    plt.title(class_names[label])
    plt.axis("off")'''

### build the model

In [ ]:
from tensorflow.keras import layers
from tensorflow.keras import regularizers

num_classes = len(class_names)

# build the model with Dropout and regulariztion



model = Sequential([
    data_augmentation,
    layers.Rescaling(1./255),
    layers.Conv2D(32, 3, padding='same', activation='relu'),
    layers.MaxPooling2D(),
    layers.Conv2D(180, 3, padding='same', activation = 'relu'),
    #layers.MaxPooling2D(),
    #layers.Conv2D(180, 3, padding = 'same', activation = 'relu'),
    layers.MaxPooling2D(),
    layers.Dropout(0.2),
    layers.Flatten(),
    layers.Dense(360, activation = 'relu'),
    #layers.Dense(num_classes)
    layers.Dense(num_classes,
                     #kernel_initializer='ones',
                     kernel_regularizer=regularizers.L1(0.01),
                     activity_regularizer=regularizers.L2(0.01))
    
])


'''
model = Sequential([
    layers.Rescaling(1./255, input_shape = (img_height, img_width, 3)),
    layers.Conv2D(16, 3, padding = 'same', activation = 'relu'),
    layers.MaxPooling2D(),
    layers.Conv2D(32, 3, padding = 'same' , activation = 'relu'),
    layers.MaxPooling2D(),
    layers.Conv2D(64, 3, padding = 'same', activation = 'relu'),
    layers.MaxPooling2D(),
    layers.Dropout(0.4),
    layers.Flatten(),
    layers.Dense(128, activation = 'relu'),
    #layers.Dense(num_classes)
    layers.Dense(num_classes,
                     kernel_initializer='ones',
                     kernel_regularizer=regularizers.L1(0.01),
                     activity_regularizer=regularizers.L2(0.01))
    
])'''

In [ ]:
model.compile(optimizer = 'adam',
             loss = tf.keras.losses.SparseCategoricalCrossentropy(from_logits = True),
             metrics = ['accuracy'])

In [ ]:
model.summary()

In [ ]:
history = model.fit(
  train_ds,
  validation_data=val_ds,
  epochs=50
)

In [ ]:
acc = history.history['accuracy']
val_acc = history.history['val_accuracy']

loss = history.history['loss']
val_loss = history.history['val_loss']

epochs_range = range(50)

plt.figure(figsize =  (8,8))
plt.subplot(1,2,1)
plt.plot(epochs_range, acc, label='Training Accuracy')
plt.plot(epochs_range, val_acc, label='Validation Accuracy')
plt.legend(loc = 'lower right')
plt.title('Training and Validation Acuuracy')

plt.subplot(1,2,2)
plt.plot(epochs_range, loss, label='Training Loss')
plt.plot(epochs_range, val_loss, label='Validation Loss')
plt.legend(loc = 'upper right')
plt.title('Training and Validation Loss')
plt.show()

### thest the following image 
### this image belongs to the creator elgeko 

In [ ]:
%ls top5creators_images/test/elgeko/

In [ ]:
img_path = 'top5creators_images/test/elgeko/QmZqYLaEtqeCdAbSjJiMPDGoirxGmb6hEoXxmSWHTApRUC.png'



# show the image
import matplotlib.image as mpimg

test_img = mpimg.imread(img_path)
plt.imshow(test_img)




In [ ]:


test_img = tf.keras.utils.load_img(img_path, target_size = (img_height, img_width))
img_array = tf.keras.utils.img_to_array(test_img)
img_array = tf.expand_dims(img_array, 0) # create the batch

predictions = model.predict(img_array)
score = tf.nn.softmax(predictions[0])


print(
"This image most likely belongs to {} with a {:.2f} percent confidence ."
.format(class_names[np.argmax(score)], 100 * np.max(score))) 

### it predict succefully the image